tests/test_aggregates.py

```diff
@@ -0,0 +1,253 @@
+
+    def test_groups_of_invalid_num(self):
+        invalid_values = [{}, 'str', -1, -1.5]
+        for invalid_value in invalid_values:
+            with pytest.raises((TypeError, ValueError)):
+                aggregator.groupsof(invalid_value)
+
+    def test_group_by_invalid_key(self):
+        with pytest.raises(TypeError):
+            aggregator.groupby(1)
+
+    def test_groupby_with_valid_type_default_for_Iterable(self):
+        aggregator.groupby(['half', 'even'], default=[-1, -1])
+
+    def test_groupby_with_invalid_type_default_key_for_Iterable(self):
+        with pytest.raises(TypeError):
+            aggregator.groupby(['half', 'even'], default=-1)
+
+    def test_groupby_with_invalid_length_default_key_for_Iterable(self):
+        with pytest.raises(ValueError):
+            aggregator.groupby(['half', 'even'], default=[-1, -1, -1])
+
+    def test_aggregate_hashing(self):
+        # Since we need to store groups on a per aggregate basis in the project,
+        # we need to be sure that the aggregates are hashing and compared correctly.
+        # This test ensures this feature.
+        def helper_default_aggregator_function(jobs):
+            return [jobs]
+
+        def helper_non_default_aggregator_function(jobs):
+            for job in jobs:
+                yield (job,)
+
+        list_of_aggregates = [
+            aggregator(), aggregator(), aggregator(helper_default_aggregator_function),
+            aggregator(helper_non_default_aggregator_function),
+            aggregator(helper_non_default_aggregator_function),
+            aggregator.groupsof(1), aggregator.groupsof(1),
+            aggregator.groupsof(2), aggregator.groupsof(3), aggregator.groupsof(4),
+            aggregator.groupby('half'), aggregator.groupby('half'), aggregator.groupby('even'),
+            aggregator.groupby('half', -1), aggregator.groupby('even', -1),
+            aggregator.groupby(['half', 'even'], default=[-1, -1])
+        ]
+
+        assert len(set(list_of_aggregates)) == 11
+
+
+# Test the _AggregatesStore and _DefaultAggregateStore classes
+class TestAggregateStoring(AggregateProjectSetup):
+
+    def test_valid_aggregator_function(self, setUp, project):
+        # Return groups of 1
+        def helper_aggregator_function(jobs):
+            for job in jobs:
+                yield (job,)
+
+        aggregate_instance = aggregator(helper_aggregator_function)
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+
+        aggregate_job_manual = helper_aggregator_function(project)
+
+        assert [aggregate for aggregate in aggregate_job_manual] == \
+               list(aggregate_instance)
+
+    def test_valid_lambda_aggregator_function(self, setUp, project):
+        aggregate_instance = aggregator(lambda jobs: [jobs])
+        aggregate_instance = aggregate_instance._create_AggregatesStore(project)
+
+        assert [tuple(project)] == list(aggregate_instance)
```

Comment:
If this is to just test with lambda's versus standard functions, we don't need both tests. If this is to test two different functions then it can be in the same test.